File: API/0418_Double_Rsi/CS/DoubleRsiStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;

using Ecng.Common;
using Ecng.Drawing;

using StockSharp.Messages;
using StockSharp.Algo;
using StockSharp.Algo.Strategies;
using StockSharp.Algo.Indicators;
using StockSharp.BusinessEntities;
using StockSharp.Localization;
using StockSharp.Charting;

namespace StockSharp.Samples.Strategies
{
	/// <summary>
	/// Double RSI Strategy with multi-timeframe analysis
	/// </summary>
	public class DoubleRsiStrategy : Strategy
	{
		private decimal _mtfRsiValue;

		public DoubleRsiStrategy()
		{
			_candleTypeParam = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
				.SetDisplay("Candle type", "Candle type for strategy calculation.", "General");

			_rsiLength = Param(nameof(RSILength), 14)
				.SetGreaterThanZero()
				.SetDisplay("RSI Length", "RSI period", "RSI");

			_mtfTimeframe = Param(nameof(MTFTimeframe), TimeSpan.FromMinutes(15).TimeFrame())
				.SetDisplay("MTF Timeframe", "Multi-timeframe for second RSI", "Multi Timeframe RSI");

			_useTP = Param(nameof(UseTP), false)
				.SetDisplay("Use Take Profit", "Enable take profit", "Take Profit");

			_tpPercent = Param(nameof(TPPercent), 1.2m)
				.SetDisplay("Take Profit %", "Take profit percentage", "Take Profit");
		}

		private readonly StrategyParam<DataType> _candleTypeParam;
		public DataType CandleType
		{
			get => _candleTypeParam.Value;
			set => _candleTypeParam.Value = value;
		}

		private readonly StrategyParam<int> _rsiLength;
		public int RSILength
		{
			get => _rsiLength.Value;
			set => _rsiLength.Value = value;
		}

		private readonly StrategyParam<DataType> _mtfTimeframe;
		public DataType MTFTimeframe
		{
			get => _mtfTimeframe.Value;
			set => _mtfTimeframe.Value = value;
		}

		private readonly StrategyParam<bool> _useTP;
		public bool UseTP
		{
			get => _useTP.Value;
			set => _useTP.Value = value;
		}

		private readonly StrategyParam<decimal> _tpPercent;
		public decimal TPPercent
		{
			get => _tpPercent.Value;
			set => _tpPercent.Value = value;
		}

		/// <inheritdoc />
		public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
			=> new[] { (Security, CandleType), (Security, MTFTimeframe) };

		/// <inheritdoc />
		protected override void OnReseted()
		{
			base.OnReseted();
			_mtfRsiValue = 0;
		}

		/// <inheritdoc />
		protected override void OnStarted(DateTimeOffset time)
		{
			base.OnStarted(time);

			// Create RSI indicators
			var rsi = new RelativeStrengthIndex { Length = RSILength };
			var mtfRsi = new RelativeStrengthIndex { Length = RSILength };

			// Subscribe to regular timeframe
			var subscription = SubscribeCandles(CandleType);
			subscription
				.Bind(rsi, OnProcessMainTimeframe)
				.Start();

			// Subscribe to multi-timeframe
			var mtfSubscription = SubscribeCandles(MTFTimeframe);
			mtfSubscription
				.Bind(mtfRsi, OnProcessMTF)
				.Start();

			// Configure chart
			var area = CreateChartArea();
			if (area != null)
			{
				DrawCandles(area, subscription);
				DrawOwnTrades(area);
			}

			// Start protection if enabled
			if (UseTP)
			{
				var takeValue = new Unit(TPPercent, UnitTypes.Percent);
				StartProtection(takeValue, new());
			}
		}

		private void OnProcessMTF(ICandleMessage candle, decimal mtfRsiValue)
		{
			if (candle.State == CandleStates.Finished)
				_mtfRsiValue = mtfRsiValue;
		}

		private void OnProcessMainTimeframe(ICandleMessage candle, decimal rsiValue)
		{
			// Only process finished candles
			if (candle.State != CandleStates.Finished)
				return;

			// Buy signal: RSI crosses above 30 and MTF RSI < 35
			var buy = rsiValue > 30 && rsiValue < 35 && _mtfRsiValue < 35;
			
			// Sell signal: RSI crosses below 70 and MTF RSI > 65
			var sell = rsiValue < 70 && rsiValue > 65 && _mtfRsiValue > 65;

			// Execute trades
			if (buy && Position <= 0)
			{
				if (Position < 0)
					RegisterOrder(this.BuyMarket(Position.Abs()));
				RegisterOrder(this.BuyMarket(Volume));
			}
			else if (sell && Position >= 0)
			{
				if (Position > 0)
					RegisterOrder(this.SellMarket(Position));
				RegisterOrder(this.SellMarket(Volume));
			}
		}
	}
}

File: API/0418_Double_Rsi/PY/double_rsi_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy


class double_rsi_strategy(Strategy):
    """Double RSI Strategy. Uses short and long RSI for entry/exit signals."""

    def __init__(self):
        super(double_rsi_strategy, self).__init__()

        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(15))) \
            .SetDisplay("Candle type", "Candle type for strategy calculation.", "General")
        self._rsi_short_length = self.Param("RSIShortLength", 7) \
            .SetDisplay("Short RSI", "Short RSI period", "RSI")
        self._rsi_long_length = self.Param("RSILongLength", 21) \
            .SetDisplay("Long RSI", "Long RSI period", "RSI")
        self._oversold = self.Param("Oversold", 35.0) \
            .SetDisplay("Oversold", "RSI oversold level", "RSI")
        self._overbought = self.Param("Overbought", 65.0) \
            .SetDisplay("Overbought", "RSI overbought level", "RSI")
        self._cooldown_bars = self.Param("CooldownBars", 10) \
            .SetDisplay("Cooldown Bars", "Bars to wait between trades", "Risk")

        self._cooldown_remaining = 0

    @property
    def CandleType(self):
        return self._candle_type.Value
    @property
    def Oversold(self):
        return self._oversold.Value
    @property
    def Overbought(self):
        return self._overbought.Value
    @property
    def CooldownBars(self):
        return self._cooldown_bars.Value

    def OnReseted(self):
        super(double_rsi_strategy, self).OnReseted()
        self._cooldown_remaining = 0

    def OnStarted(self, time):
        super(double_rsi_strategy, self).OnStarted(time)
        rsi_short = RelativeStrengthIndex()
        rsi_short.Length = self._rsi_short_length.Value
        rsi_long = RelativeStrengthIndex()
        rsi_long.Length = self._rsi_long_length.Value
        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(rsi_short, rsi_long, self.OnProcess).Start()

    def OnProcess(self, candle, rsi_short_val, rsi_long_val):
        if candle.State != CandleStates.Finished:
            return
        if not self.IsFormedAndOnlineAndAllowTrading():
            return
        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1
            return

        rs = float(rsi_short_val)
        rl = float(rsi_long_val)

        if rs < self.Oversold and rl < self.Oversold and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
            self._cooldown_remaining = self.CooldownBars
        elif rs > self.Overbought and rl > self.Overbought and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()
            self._cooldown_remaining = self.CooldownBars
        elif self.Position > 0 and rs > self.Overbought:
            self.SellMarket()
            self._cooldown_remaining = self.CooldownBars
        elif self.Position < 0 and rs < self.Oversold:
            self.BuyMarket()
            self._cooldown_remaining = self.CooldownBars

    def CreateClone(self):
        return double_rsi_strategy()
